In [50]:
import ant_colony


In [51]:
from ant_colony.aco_tsp import *

In [52]:
from ant_colony.utils import *
from ant_colony.aco_tsp_oo import *

In [53]:
# !pip install line-profiler
# !pip install memory_profiler

In [54]:
import timeit
import cProfile
import pstats
import memory_profiler
from memory_profiler import memory_usage

In [ ]:
## import line_profiler
NO FUNCIONa

# En primer lugar, veamos las características de la máquina que se utilizarán para el perfilamiento: 

In [40]:
# Esto deberia funcionar en linux

#%%bash
#lscpu

In [41]:
#%%bash
#sudo lshw -C memory

In [55]:
path = '../datasets/gr17_d_city_distances.txt'

In [56]:
G = read_data(path)


In [10]:
G

# Primero vamos a medir los tiempos de ejecucion

## Timeit para medir tiempos de ejecución

In [16]:
colonia = colony(G, 
                 init_node=7, 
                 n_ants=5, 
                 max_iter=500, 
                 rho=.5, 
                 alpha=3,
                 beta=3)

## Probamos el metodo solve.tsp

### Primero medimos el tiempo que toma ejecutar el método 

In [18]:
start = timeit.timeit()
colonia.solve_tsp()
end = timeit.timeit()
print(f"- Tiempo total: {(end-start)} segundos.")

- Tiempo total: -0.0028420799999366864 segundos.


### A continuación utilizamos cProfile para ver las partes del método que más se tardan 

In [23]:
cprof = cProfile.Profile()
cprof.enable()
res =  colonia.solve_tsp()
cprof.disable()
cprof.print_stats(sort='cumtime')

         3141046 function calls (3140546 primitive calls) in 3.035 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    3.035    1.517 interactiveshell.py:3376(run_code)
        2    0.000    0.000    3.035    1.517 {built-in method builtins.exec}
        1    0.000    0.000    3.034    3.034 <ipython-input-23-5fb9737eed66>:3(<module>)
        1    0.015    0.015    3.034    3.034 aco_tsp_oo.py:112(solve_tsp)
      500    0.077    0.000    2.822    0.006 aco_tsp_oo.py:81(_colony_run)
     2500    0.309    0.000    2.739    0.001 aco_tsp_oo.py:151(walk_over_graph)
     2500    0.351    0.000    2.071    0.001 utils.py:25(create_dic_dist_from_graph)
   767500    0.872    0.000    1.112    0.000 defmatrix.py:189(__getitem__)
     2500    0.012    0.000    0.602    0.000 convert_matrix.py:442(to_numpy_matrix)
     2500    0.391    0.000    0.539    0.000 convert_matrix.py:1093(to_numpy_array)
    400

Y observamos que lo que mas se tarda es : 
    
     2500    0.309    0.000    2.739    0.001 aco_tsp_oo.py:151(walk_over_graph)
     2500    0.351    0.000    2.071    0.001 utils.py:25(create_dic_dist_from_graph)
   767500    0.872    0.000    1.112    0.000 defmatrix.py:189(__getitem__)

In [24]:
cprof.dump_stats("solve_tsp_stats")


In [27]:
p_solve_tsp_stats = pstats.Stats("solve_tsp_stats")
print(p_solve_tsp_stats.sort_stats("cumulative").print_stats(10))

Wed Apr 28 14:09:48 2021    solve_tsp_stats

         3141046 function calls (3140546 primitive calls) in 3.035 seconds

   Ordered by: cumulative time
   List reduced from 69 to 10 due to restriction <10>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000    3.035    1.517 /Users/sansansansan/.pyenv/versions/3.7.4/envs/itam_intro_to_ds/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3376(run_code)
        2    0.000    0.000    3.035    1.517 {built-in method builtins.exec}
        1    0.000    0.000    3.034    3.034 <ipython-input-23-5fb9737eed66>:3(<module>)
        1    0.015    0.015    3.034    3.034 /Users/sansansansan/.pyenv/versions/3.7.4/envs/itam_intro_to_ds/lib/python3.7/site-packages/ant_colony/aco_tsp_oo.py:112(solve_tsp)
      500    0.077    0.000    2.822    0.006 /Users/sansansansan/.pyenv/versions/3.7.4/envs/itam_intro_to_ds/lib/python3.7/site-packages/ant_colony/aco_tsp_oo.py:81(_colony_run)
     2500 

### Nuevamente observamos que walk over graph , get item y create dict dist from graph son las que mas se tardan 



# Ahora realizamos el perfilamiento con respecto a la memoria

In [49]:
t = (colonia.solve_tsp())
print(memory_usage(t, max_usage=True))

KeyboardInterrupt: 

In [ ]:
start_mem = memory_usage(max_usage=True)
res = memory_usage(t, max_usage=True, retval=True)
print('start mem', start_mem)
print('max mem', res[0])
print('used mem', res[0]-start_mem)
print('fun output', res[1])

In [57]:
%load_ext memory_profiler


In [58]:
%memit #how much RAM this process is consuming


peak memory: 128.07 MiB, increment: 0.39 MiB


In [59]:
%memit -c colonia.solve_tsp()


peak memory: 134.00 MiB, increment: 6.74 MiB
